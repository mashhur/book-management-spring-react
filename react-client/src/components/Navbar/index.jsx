import react from "react";
import { Link } from "react-router-dom";
import { TiShoppingCart } from "react-icons/ti";
import { useSelector } from "react-redux";
function Navbar() {
	const user = useSelector((state) => state.user);
	return (
		<nav className="navbar navbar-expand-lg border-bottom  navbar-light bg-light">
			<div className="container-fluid">
				<Link className="navbar-brand" to="/">
					<p
						style={{ opactity: ".5" }}
						className="fs-2 mb-0 ">
						<span style={{ letterSpacing: "-1px" }}>
							Diamond
						</span>
						<small
							style={{
								fontSize: "18px",
								letterSpacing: "2px",
							}}>
							<b>BOOKS</b>
						</small>
					</p>
				</Link>
				<button
					className="navbar-toggler"
					type="button"
					data-toggle="collapse"
					data-target="#navbarNav"
					aria-controls="navbarNav"
					aria-expanded="false"
					aria-label="Toggle navigation">
					<span className="navbar-toggler-icon"></span>
				</button>
				<div className="w-100 d-flex justify-content-center">
					<input
						className="form-search  px-4 w-50 mx-5"
						type="search"
						placeholder="Search for a book, author, or category"
						aria-label="Search"
					/>
				</div>
				<div
					className="collapse navbar-collapse"
					id="navbarNav">
					<ul className="navbar-nav ms-auto">
						<li className="nav-item ">
							<Link className="nav-link" to="/cart">
								<div className="position-relative">
									<TiShoppingCart
										style={{
											fontSize: "40px",
											opacity: 0.8,
										}}
									/>
									{user.cart.length > 0 && (
										<span
											data-test="cart-item-count"
											className="position-absolute top-0 start-100 translate-middle badge rounded-pill bg-success">
											{user.cart.length}
										</span>
									)}
								</div>
							</Link>
						</li>
						<li className="nav-item active">
							<Link className="nav-link" to="/profile">
								<img
									style={{ height: "45px", width: "45px" }}
									src="https://www.w3schools.com/howto/img_avatar2.png"
									className="rounded-circle"
									alt="avatar"
								/>
							</Link>
						</li>
					</ul>
				</div>
			</div>
		</nav>
	);
}
export default Navbar;
