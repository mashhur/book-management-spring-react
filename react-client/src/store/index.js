import { configureStore } from "@reduxjs/toolkit";
import locationReducer from "./slices/locationSlice";
import userReducer from "./slices/userSlice";

export const store = configureStore({
	reducer: {
		location: locationReducer,
		user: userReducer,
	},
});
