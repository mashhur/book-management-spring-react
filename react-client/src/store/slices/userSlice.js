import { createSlice } from "@reduxjs/toolkit";

const inituser = JSON.parse(localStorage.getItem("user"));
const user = inituser || {
	cid: "",
	fullName: "",
	phoneNumber: "",
};
const initialState = {
	auth: localStorage.getItem("auth") || false,
	user: user,
	address: {
		aid: "",
		customerName: "",
		phoneNumber: "",
		address: "",
		mstate: "",
		pincode: "",
		customerId: "",
	},
	books: [],
	cart: [],
};

export const userSlice = createSlice({
	name: "user",
	initialState,
	reducers: {
		changeAuth: (state) => {
			state.auth = true;
		},
		setUser: (state, action) => {
			state.user = action.payload;
		},
		setAddress(state, action) {
			state.address = action.payload;
		},
		addBook(state, action) {
			state.books = [];
			state.books.push(action.payload);
		},
		addBooktoCart(state, action) {
			state.cart.push(action.payload);
		},
		addtoBuy(state) {
			const cart = state.cart;
			state.books = cart;
			state.cart = [];
		},
		clearOrderDetails(state) {
			state.books = [];
			state.address = {};
		},
		removeBook(state, action) {
			state.cart.splice(action.payload, 1);
		},
		clearAuth(state) {
			state.auth = false;
			state.user = {
				cid: "",
				fullName: "",
				phoneNumber: "",
			};
		},
	},
});

export const {
	changeAuth,
	setUser,
	setAddress,
	addBook,
	clearOrderDetails,
	removeBook,
	addBooktoCart,
	addtoBuy,
	clearAuth,
} = userSlice.actions;

export default userSlice.reducer;
