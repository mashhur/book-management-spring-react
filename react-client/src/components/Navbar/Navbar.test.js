import { mount } from "enzyme";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import Navbar from ".";
import { createTestStore } from "../../store";
import { addBooktoCart } from "../../store/slices/userSlice";
import { findByTestAttr } from "../../tests/testUtils";

const bookStore = {
	books: [
		{
			bid: 1,
			title: "",
			price: "",
			description: "",
			cover: "",
		},
	],
};

const store = createTestStore();
const setup = () => {
	return mount(
		<BrowserRouter>
			<Provider store={store}>
				<Navbar />
			</Provider>
		</BrowserRouter>
	);
};

it("when cart is empty renders nothing", async () => {
	const wrapper = setup();
	const component = await findByTestAttr(
		wrapper,
		"cart-item-count"
	);
	expect(component.length).toBe(0);
});
it("when cart is filled renders count", async () => {
	store.dispatch(addBooktoCart(bookStore.books[0]));
	const wrapper = setup();
	const component = await findByTestAttr(
		wrapper,
		"cart-item-count"
	);
	expect(component.text()).toBe("1");
});
