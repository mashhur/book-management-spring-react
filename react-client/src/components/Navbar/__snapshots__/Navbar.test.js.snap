// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the deafult snapshot 1`] = `ReactWrapper {}`;

exports[`should match the error snapshot 1`] = `ReactWrapper {}`;
